File: recursion/day2/binarySearch.java

// package patterns.recursion.day2;
import java.util.*;

public class binarySearch {
    public static void main(String[] args) {
        System.out.println("Enter no of elements in the array: ");
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int arr[] = new int[n];
        System.out.println("Enetr elements of the array: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        System.out.println("Enter the key: ");
        int key = sc.nextInt();
        int ans = findKey(arr, 0, n, key);
        if (ans < 0) {
            System.out.println(key + " is not found");
        } else {
            System.out.println(key + " is found at index " + (ans + 1));
        }
    }

    public static int findKey(int arr[], int m, int n, int key) {
        int s = m, e = n - 1;
        if (s > e) {
            return -1;
        }
        int mid = s + (e - s) / 2;
        if (arr[mid] == key)
            return mid;
        if (arr[mid] < key) {
            return findKey(arr, mid + 1, e, key);
        } else {
            return findKey(arr, s, mid - 1, key);
        }
    }
}

File: practice/array/binary search.cpp
#include <bits/stdc++.h>
using namespace std;

int binary_search(int arr[], int n, int key)
{
    int s = 0, e = n - 1;

    // int mid = (s + e) / 2;
    while (s <= e)
    {
        int mid = s + (e - s) / 2;
        if (arr[mid] == key)
        {
            return mid;
        }
        if (key < arr[mid])
        {
            e = mid - 1;
        }
        else
        {
            s = mid + 1;
        }
        // mid = s + (e - s) / 2;
        // mid = (s + e) / 2;
    }
    return -1;
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6};
    int key = 4;
    int ans = binary_search(arr, 6, key);
    if (ans > 0)
    {
        cout << "Index: " << ans << endl;
    }
    else
    {
        cout << key << " is not found" << endl;
    }
}